controller/login.go

```diff
@@ -1,48 +1,54 @@
 package controller
 
 import (
-	"fmt"
 	"github.com/fabric8-services/fabric8-auth/app"
-	"github.com/fabric8-services/fabric8-auth/configuration"
-	"github.com/fabric8-services/fabric8-auth/login"
+	"github.com/fabric8-services/fabric8-auth/application"
+	"github.com/fabric8-services/fabric8-auth/client"
+	"github.com/fabric8-services/fabric8-auth/jsonapi"
 	"github.com/fabric8-services/fabric8-auth/rest"
-	"github.com/fabric8-services/fabric8-auth/token"
-
-	client "github.com/fabric8-services/fabric8-auth/client"
 	"github.com/goadesign/goa"
+	"github.com/satori/go.uuid"
 )
 
-type LoginConfiguration interface {
-	login.Configuration
-	GetKeycloakEndpointAuth(*goa.RequestData) (string, error)
-	GetKeycloakURL() string
-	GetKeycloakRealm() string
-	GetPublicOauthClientID() string
-	GetServiceAccounts() map[string]configuration.ServiceAccount
-}
-
 // LoginController implements the login resource.
 type LoginController struct {
 	*goa.Controller
-	Auth          login.KeycloakOAuthService
-	TokenManager  token.Manager
-	Configuration LoginConfiguration
+	app application.Application
 }
 
 // NewLoginController creates a login controller.
-func NewLoginController(service *goa.Service, auth *login.KeycloakOAuthProvider, tokenManager token.Manager, configuration LoginConfiguration) *LoginController {
-	return &LoginController{Controller: service.NewController("login"), Auth: auth, TokenManager: tokenManager, Configuration: configuration}
+func NewLoginController(service *goa.Service, app application.Application) *LoginController {
+	return &LoginController{
+		Controller: service.NewController("login"),
+		app:        app}
 }
 
 // Login runs the login action.
 func (c *LoginController) Login(ctx *app.LoginLoginContext) error {
+	// Generate a new unique state value
+	state := uuid.NewV4().String()
+
+	// Get the URL of the callback endpoint, the client will be redirected here after being redirected to the authentication provider
+	callbackURL := rest.AbsoluteURL(ctx.RequestData, client.CallbackLoginPath(), nil)
 
-	oauthIdentityProvider := login.NewIdentityProvider(c.Configuration)
-	oauthIdentityProvider.RedirectURL = rest.AbsoluteURL(ctx.RequestData, client.LoginLoginPath(), nil)
-	if ctx.Scope != nil {
-		oauthIdentityProvider.Endpoint.AuthURL = fmt.Sprintf("%s?scope=%s", oauthIdentityProvider.Endpoint.AuthURL, *ctx.Scope) // Offline token
+	redirectURL, err := c.app.AuthenticationProviderService().GenerateAuthCodeURL(ctx, ctx.Redirect, ctx.APIClient,
```

Comment:
- When `/api/login?redirect=xyz` is called, we return the AuthCodeURL which asks the user to authenticate on RHD.
- After signing in, `/api/login?state=xyz&code=abc` is called.

I'm guess you are yet to add support for the second part?

Reply:
I've refactored the second part into a separate callback and method, see `AuthenticationProviderService.LoginCallback()`.

Reply:
But it would be called from the same controller right, because clients would be expecting so ? `/api/login`

Reply:
The controller has a new `Callback()` endpoint, `/api/login/callback`.

Reply:
Right, your are doing this 
```
// Get the URL of the callback endpoint, the client will be redirected here after being redirected to the authentication provider
callbackURL := rest.AbsoluteURL(ctx.RequestData, client.CallbackAuthorizePath(), nil)
```